common/autoware_universe_utils/src/geometry/geometry.cpp

```diff
@@ -390,4 +390,425 @@ bool intersects_convex(const Polygon2d & convex_polygon1, const Polygon2d & conv
+        winding_number--;
+        continue;
+      }
+    } else {
+      continue;
+    }
+
+    if (std::abs(cross) <= std::numeric_limits<double>::epsilon()) {  // point is on edge
+      return true;
+    }
+  }
+
+  return winding_number != 0;
+}
+
+bool disjoint(const alt::ConvexPolygon2d & poly1, const alt::ConvexPolygon2d & poly2)
+{
+  if (equals(poly1, poly2)) {
+    return false;
+  }
+
+  if (intersects(poly1, poly2)) {
+    return false;
+  }
+
+  for (const auto & vertex : poly1.vertices()) {
+    if (touches(vertex, poly2)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+double distance(
+  const alt::Point2d & point, const alt::Point2d & seg_start, const alt::Point2d & seg_end)
+{
+  const auto seg_vec = seg_end - seg_start;
+  const auto point_vec = point - seg_start;
+
+  const double seg_vec_norm = seg_vec.norm();
+  const double seg_point_dot = seg_vec.dot(point_vec);
+
+  if (seg_vec_norm <= std::numeric_limits<double>::epsilon() || seg_point_dot < 0) {
+    return point_vec.norm();
+  } else if (seg_point_dot > std::pow(seg_vec_norm, 2)) {
+    return (point - seg_end).norm();
+  } else {
+    return std::abs(seg_vec.cross(point_vec)) / seg_vec_norm;
+  }
+}
+
+double distance(const alt::Point2d & point, const alt::ConvexPolygon2d & poly)
+{
+  if (covered_by(point, poly)) {
+    return 0.0;
+  }
+
+  // TODO(mitukou1109): Use plane sweep method to improve performance?
+  const auto & vertices = poly.vertices();
+  double min_distance = std::numeric_limits<double>::max();
+  for (size_t i = 0; i < vertices.size(); ++i) {
+    min_distance =
+      std::min(min_distance, distance(point, vertices[i], vertices[(i + 1) % vertices.size()]));
+  }
+
+  return min_distance;
+}
+
+bool equals(const alt::Point2d & point1, const alt::Point2d & point2)
+{
+  return std::abs(point1.x() - point2.x()) <= std::numeric_limits<double>::epsilon() &&
+         std::abs(point1.y() - point2.y()) <= std::numeric_limits<double>::epsilon();
+}
+
+bool equals(const alt::ConvexPolygon2d & poly1, const alt::ConvexPolygon2d & poly2)
+{
+  return std::all_of(poly1.vertices().begin(), poly1.vertices().end(), [&](const auto & a) {
+    return std::any_of(poly2.vertices().begin(), poly2.vertices().end(), [&](const auto & b) {
+      return equals(a, b);
+    });
+  });
+}
+
+bool intersects(
+  const alt::Point2d & seg1_start, const alt::Point2d & seg1_end, const alt::Point2d & seg2_start,
+  const alt::Point2d & seg2_end)
+{
+  const auto v1 = seg1_end - seg1_start;
+  const auto v2 = seg2_end - seg2_start;
+
+  const auto det = v1.cross(v2);
+  if (std::abs(det) <= std::numeric_limits<double>::epsilon()) {
+    return false;
+  }
+
+  const auto v12 = seg2_end - seg1_end;
+  const double t = v2.cross(v12) / det;
+  const double s = v1.cross(v12) / det;
+  if (t < 0 || 1 < t || s < 0 || 1 < s) {
+    return false;
+  }
+
+  return true;
+}
+
+bool intersects(const alt::ConvexPolygon2d & poly1, const alt::ConvexPolygon2d & poly2)
```

Comment:
GJK is already implemented (https://github.com/autowarefoundation/autoware.universe/blob/main/common/autoware_universe_utils/src/geometry/gjk_2d.cpp) but it only works for convex polygon.
If your goal is to provide an alternative to the `boost::geometry::intersects` function, then you need to use a more general algorithm that also works for concave planners.

Reply:
I agree, but I already named the class for representing polygons `ConvexPolygon2d` and thus I do not expect the input to be concave (at least for now).